File: content/manual/v8.2/source/includes/csfle/java/src/main/java/com/mongodb/csfle/InsertEncryptedDocument.java
package com.mongodb.csfle;

import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

import static com.mongodb.client.model.Filters.eq;

import com.mongodb.AutoEncryptionSettings;
import com.mongodb.ConnectionString;
import com.mongodb.MongoClientSettings;
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoClients;
import org.bson.BsonDocument;
import org.bson.Document;

public class InsertEncryptedDocument {

    public static void main(String[] args) throws Exception {
        Map<String, Map<String, Object>> kmsProviders =
            Config.getKmsProviders();

        // start-json-schema
        String dekId = new String(
            Files.readAllBytes(Paths.get(Config.DEK_ID_PATH))).trim();
        Document jsonSchema = new Document()
            .append("bsonType", "object")
            .append("encryptMetadata", new Document()
                .append("keyId", new ArrayList<>(Arrays.asList(
                    new Document()
                        .append("$binary", new Document()
                            .append("base64", dekId)
                            .append("subType", "04"))))))
            .append("properties", new Document()
                .append("ssn", new Document()
                    .append("encrypt", new Document()
                        .append("bsonType", "int")
                        .append("algorithm",
                            "AEAD_AES_256_CBC_HMAC_SHA_512-Deterministic")))
                .append("bloodType", new Document()
                    .append("encrypt", new Document()
                        .append("bsonType", "string")
                        .append("algorithm",
                            "AEAD_AES_256_CBC_HMAC_SHA_512-Random")))
                .append("medicalRecords", new Document()
                    .append("encrypt", new Document()
                        .append("bsonType", "array")
                        .append("algorithm",
                            "AEAD_AES_256_CBC_HMAC_SHA_512-Random")))
                .append("insurance", new Document()
                    .append("bsonType", "object")
                    .append("properties", new Document()
                        .append("policyNumber", new Document()
                            .append("encrypt", new Document()
                                .append("bsonType", "int")
                                .append("algorithm",
                                    "AEAD_AES_256_CBC_HMAC_SHA_512-Deterministic"))))));
        Map<String, BsonDocument> schemaMap = new HashMap<>();
        schemaMap.put("medicalRecords.patients",
            BsonDocument.parse(jsonSchema.toJson()));
        // end-json-schema

        // start-create-client
        Map<String, Object> extraOptions = new HashMap<>();
        extraOptions.put("cryptSharedLibPath", Config.CRYPT_SHARED_LIB_PATH);
        MongoClientSettings clientSettings = MongoClientSettings.builder()
            .applyConnectionString(
                new ConnectionString(Config.CONNECTION_STRING))
            .autoEncryptionSettings(AutoEncryptionSettings.builder()
                .keyVaultNamespace(Config.KEY_VAULT_NAMESPACE)
                .kmsProviders(kmsProviders)
                .schemaMap(schemaMap)
                .extraOptions(extraOptions)
                .build())
            .build();
        MongoClient mongoClientRegular = MongoClients.create(Config.CONNECTION_STRING);
        MongoClient mongoClientSecure = MongoClients.create(clientSettings);
        // end-create-client

        // start-insert-document
        ArrayList<Document> medicalRecords = new ArrayList<>();
        medicalRecords.add(new Document().append("weight", 180));
        medicalRecords.add(
            new Document().append("bloodPressure", "120/80"));
        Document insurance = new Document()
            .append("policyNumber", 123142)
            .append("provider", "MaestCare");
        Document patient = new Document()
            .append("name", "Jon Doe")
            .append("ssn", 241014209)
            .append("bloodType", "AB+")
            .append("medicalRecords", medicalRecords)
            .append("insurance", insurance);
        mongoClientSecure.getDatabase("medicalRecords")
            .getCollection("patients").insertOne(patient);
        // end-insert-document

        // start-find-document
        System.out.println(
            "Finding a document with the regular (non-encrypted) client:");
        Document docRegular = mongoClientRegular
            .getDatabase("medicalRecords").getCollection("patients")
            .find(eq("name", "Jon Doe")).first();
        System.out.println(docRegular.toJson());
        System.out.println(
            "Finding a document with the encrypted client:");
        Document docSecure = mongoClientSecure
            .getDatabase("medicalRecords").getCollection("patients")
            .find(eq("name", "Jon Doe")).first();
        System.out.println(docSecure.toJson());
        // end-find-document

        mongoClientSecure.close();
        mongoClientRegular.close();
    }
}

File: content/manual/manual/source/includes/csfle/csharp/InsertEncryptedDocument.cs
using System;
using System.IO;
using System.Collections.Generic;
using MongoDB.Driver;
using MongoDB.Bson;
using MongoDB.Driver.Encryption;

namespace CsfleTutorial;

public static class InsertEncryptedDocument
{
    public static void Insert()
    {
        var db = "medicalRecords";
        var coll = "patients";
        var dbNamespace = $"{db}.{coll}";

        // start-json-schema
        var dekId = File.ReadAllText(Config.DekIdPath);
        var schema = new BsonDocument
        {
            { "bsonType", "object" },
            {
                "encryptMetadata",
                new BsonDocument("keyId", new BsonArray(new[]
                {
                    new BsonBinaryData(
                        Convert.FromBase64String(dekId),
                        BsonBinarySubType.UuidStandard)
                }))
            },
            {
                "properties", new BsonDocument
                {
                    {
                        "ssn", new BsonDocument
                        {
                            {
                                "encrypt", new BsonDocument
                                {
                                    { "bsonType", "int" },
                                    {
                                        "algorithm",
                                        "AEAD_AES_256_CBC_HMAC_SHA_512-Deterministic"
                                    }
                                }
                            }
                        }
                    },
                    {
                        "bloodType", new BsonDocument
                        {
                            {
                                "encrypt", new BsonDocument
                                {
                                    { "bsonType", "string" },
                                    {
                                        "algorithm",
                                        "AEAD_AES_256_CBC_HMAC_SHA_512-Random"
                                    }
                                }
                            }
                        }
                    },
                    {
                        "medicalRecords", new BsonDocument
                        {
                            {
                                "encrypt", new BsonDocument
                                {
                                    { "bsonType", "array" },
                                    {
                                        "algorithm",
                                        "AEAD_AES_256_CBC_HMAC_SHA_512-Random"
                                    }
                                }
                            }
                        }
                    },
                    {
                        "insurance", new BsonDocument
                        {
                            { "bsonType", "object" },
                            {
                                "properties", new BsonDocument
                                {
                                    {
                                        "policyNumber", new BsonDocument
                                        {
                                            {
                                                "encrypt", new BsonDocument
                                                {
                                                    { "bsonType", "int" },
                                                    {
                                                        "algorithm",
                                                        "AEAD_AES_256_CBC_HMAC_SHA_512-Deterministic"
                                                    }
                                                }
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        };
        var schemaMap = new Dictionary<string, BsonDocument>
        {
            { dbNamespace, schema }
        };
        // end-json-schema

        // start-create-client
        var keyVaultNamespace =
            CollectionNamespace.FromFullName(Config.KeyVaultNamespace);
        var kmsProviders = Config.GetKmsProviders();
        var extraOptions = new Dictionary<string, object>
        {
            { "cryptSharedLibPath", Config.CryptSharedLibPath }
        };
        MongoClientSettings.Extensions.AddAutoEncryption();
        var clientSettings =
            MongoClientSettings.FromConnectionString(Config.ConnectionString);
        var autoEncryptionOptions = new AutoEncryptionOptions(
            keyVaultNamespace: keyVaultNamespace,
            kmsProviders: kmsProviders,
            schemaMap: schemaMap,
            extraOptions: extraOptions);
        clientSettings.AutoEncryptionOptions = autoEncryptionOptions;
        var secureClient = new MongoClient(clientSettings);
        var regularClient = new MongoClient(Config.ConnectionString);
        // end-create-client

        // start-insert-document
        var secureCollection = secureClient
            .GetDatabase(db).GetCollection<BsonDocument>(coll);
        var sampleDoc = new BsonDocument
        {
            { "name", "Jon Doe" },
            { "ssn", 241014209 },
            { "bloodType", "AB+" },
            {
                "medicalRecords", new BsonArray
                {
                    new BsonDocument
                    {
                        { "weight", 180 },
                        { "bloodPressure", "120/80" }
                    }
                }
            },
            {
                "insurance", new BsonDocument
                {
                    { "policyNumber", 123142 },
                    { "provider", "MaestCare" }
                }
            }
        };
        secureCollection.InsertOne(sampleDoc);
        // end-insert-document

        // start-find-document
        var regularCollection = regularClient
            .GetDatabase(db).GetCollection<BsonDocument>(coll);

        Console.WriteLine(
            "Finding a document with the regular (non-encrypted) client:");
        var filter = Builders<BsonDocument>.Filter.Eq("name", "Jon Doe");
        var regularResult = regularCollection.Find(filter).First();
        Console.WriteLine($"\n{regularResult}\n");

        Console.WriteLine("Finding a document with the encrypted client:");
        var secureResult = secureCollection.Find(filter).First();
        Console.WriteLine($"\n{secureResult}\n");
        // end-find-document
    }
}
